unit-test/background/abn-framework.js

```diff
@@ -125,3 +160,191 @@ describe('choseCohort', () => {
+    let pixelIntercept = null;
+
+    beforeEach(() => {
+        pixelRequests = [];
+        pixelIntercept = spyOn(load, 'url').and.callFake((url) => {
+            const parsed = new URL(url);
+            pixelRequests.push(parsed.pathname + parsed.search);
+        });
+    });
+
+    it('markExperimentEnrolled sends a correct pixel only once', () => {
+        const { remoteConfig, abnMetrics } = constructMockComponents();
+        remoteConfig.updateConfig(mockExperimentConfig);
+        abnMetrics.markExperimentEnrolled(feature, subFeature);
+        expect(pixelIntercept).toHaveBeenCalledTimes(1);
+        expect(pixelValidator.validateLivePixels(experimentPixels['experiment.enroll'], 'experiment.enroll', pixelRequests[0])).toEqual([]);
+
+        // call a second time: pixel shouldn't be triggered again
+        abnMetrics.markExperimentEnrolled(feature, subFeature);
+        expect(pixelIntercept).toHaveBeenCalledTimes(1);
+    });
+
+    it('markExperimentEnrolled with no metrics creates the default set', () => {
+        const { remoteConfig, abnMetrics } = constructMockComponents();
+        remoteConfig.updateConfig(mockExperimentConfig);
+        abnMetrics.markExperimentEnrolled(feature, subFeature);
+        expect(remoteConfig.getCohort(feature, subFeature).metrics.length).toEqual(38);
+        expect(pixelIntercept).toHaveBeenCalledTimes(1);
+    });
+
+    it('markExperimentEnrolled sends no pixel if experiment does not exist', () => {
+        const { remoteConfig, abnMetrics } = constructMockComponents();
+        remoteConfig.updateConfig(mockExperimentConfig);
+        abnMetrics.markExperimentEnrolled(feature, 'fooFeature2');
+        expect(pixelIntercept).toHaveBeenCalledTimes(0);
+    });
+
+    it('onMetricTriggered triggers a metric pixel exactly once for an experiment', () => {
+        const { remoteConfig, abnMetrics } = constructMockComponents();
+        remoteConfig.updateConfig(mockExperimentConfig);
+        abnMetrics.markExperimentEnrolled(feature, subFeature);
+        abnMetrics.onMetricTriggered('search');
+        abnMetrics.onMetricTriggered('search');
+        // enrollment + metric pixels
+        expect(pixelIntercept).toHaveBeenCalledTimes(2);
+        expect(pixelRequests[1]).toContain('conversionWindowDays=0&');
+        expect(pixelRequests[1]).toContain('value=1&');
+        expect(pixelValidator.validateLivePixels(experimentPixels['experiment.metrics'], 'experiment.metrics', pixelRequests[1])).toEqual(
+            [],
+        );
+    });
+
+    it('onMetricTriggered can trigger multiple matching metrics', () => {
+        const { remoteConfig, abnMetrics } = constructMockComponents();
+        remoteConfig.updateConfig(mockExperimentConfig);
+        abnMetrics.markExperimentEnrolled(feature, subFeature);
+        // modify enrollment to be 6 days ago
+        const cohort = remoteConfig.getCohort(feature, subFeature);
+        cohort.enrolledAt = Date.now() - 1000 * 60 * 60 * 25 * 6;
```

Comment:
Is `Date.now() - 1000 * 60 * 60 * 25 * 6;` a typo? (I figure it should be `* 25`, as in 24 hours in a day.) Same elsewhere, also maybe have a `millisecondsInDay` or similar constant in this file to reuse?

Reply:
No - I used 25 hours per day so that the events land safely within a time window. Using 24 hours would make it potentially unclear which bucket the metric should fall into (i.e. event happens _exactly_ 24 hours after enrollment, is that day 0 or day 1?). I've added a brief comment to say why 25 hour is used.